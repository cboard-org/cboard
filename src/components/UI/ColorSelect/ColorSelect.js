import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import CloseIcon from '@material-ui/icons/Close';

import IconButton from '../IconButton';
import Circle from './Circle';
import messages from './ColorSelect.messages';

const propTypes = {
  colors: PropTypes.arrayOf(PropTypes.string),
  intl: intlShape.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedColor: PropTypes.string
};

const defaultProps = {
  colors: ['#CE93D8', '#2196F3', '#4CAF50', '#E57373']
};

const ColorSelect = props => {
  const { colors, intl, onChange, selectedColor } = props;

  const colorLabel = intl.formatMessage(messages.color);
  const radioGroupStyle = { flexDirection: 'row' };
  const circleStrokeWidth = 2;

  return (
    <FormControl className="ColorSelect">
      <FormLabel>{colorLabel}</FormLabel>
      <RadioGroup
        aria-label={colorLabel}
        name="color"
        value={selectedColor}
        style={radioGroupStyle}
        onChange={onChange}
      >
        {colors.map(color => (
          <Radio
            key={color}
            value={color}
            icon={<Circle fill={color} />}
            checkedIcon={
              <Circle fill={color} strokeWidth={circleStrokeWidth} />
            }
          />
        ))}
        {selectedColor && (
          <IconButton
            label={intl.formatMessage(messages.clearSelection)}
            onClick={() => {
              onChange();
            }}
          >
            <CloseIcon />
          </IconButton>
        )}
      </RadioGroup>
    </FormControl>
  );
};

ColorSelect.propTypes = propTypes;
ColorSelect.defaultProps = defaultProps;

export default injectIntl(ColorSelect);
