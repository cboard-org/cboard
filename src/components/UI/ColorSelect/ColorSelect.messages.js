import { defineMessages } from 'react-intl';

export default defineMessages({
  color: {
    id: 'cboard.components.ColorSelect.color',
    defaultMessage: 'Color'
  },
  clearSelection: {
    id: 'cboard.components.ColorSelect.clearSelection',
    defaultMessage: 'Clear selection'
  }
});
