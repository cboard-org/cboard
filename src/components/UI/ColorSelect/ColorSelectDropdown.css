.colorSelectDropdown {
  display: flex;
  flex-direction: column;
}

.MuiFormControl-root.colorSelectDropdown-options {
  position: absolute;
  margin-top: 4em;
  margin-left: -1em;
}

.closed {
  display: none;
}
