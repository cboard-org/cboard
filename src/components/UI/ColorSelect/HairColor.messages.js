import { defineMessages } from 'react-intl';

export default defineMessages({
  skinTone: {
    id: 'cboard.components.HairColor.hairColor',
    defaultMessage: 'Hair Color'
  },
  clearSelection: {
    id: 'cboard.components.HairColor.clearSelection',
    defaultMessage: 'Clear selection'
  }
});
