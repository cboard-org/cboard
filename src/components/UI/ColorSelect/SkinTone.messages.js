import { defineMessages } from 'react-intl';

export default defineMessages({
  skinTone: {
    id: 'cboard.components.SkinTone.skinTone',
    defaultMessage: 'Skin Tone'
  },
  clearSelection: {
    id: 'cboard.components.SkinTone.clearSelection',
    defaultMessage: 'Clear selection'
  }
});
