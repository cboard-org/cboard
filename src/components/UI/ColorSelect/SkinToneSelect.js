import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FormControl from '@material-ui/core/FormControl';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import CloseIcon from '@material-ui/icons/Close';

import IconButton from '../IconButton';
import Circle from './Circle';
import messages from './SkinTone.messages';

const skinToneSources = new Map([
  [
    'arasaac',
    [
      {
        name: 'white',
        color: '#F5E5DE'
      },
      {
        name: 'black',
        color: '#A65C17'
      },
      {
        name: 'assian',
        color: '#F4ECAD'
      },
      {
        name: 'mulatto',
        color: '#E3AB72'
      },
      {
        name: 'aztec',
        color: '#CF9D7C'
      }
    ]
  ]
]);

const propTypes = {
  source: PropTypes.string.isRequired,
  intl: intlShape.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedColor: PropTypes.string.isRequired
};

class SkinToneSelect extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      skinToneMenu: skinToneSources.has(props.source)
        ? skinToneSources.get(props.source)
        : skinToneSources.get('arasaac')
    };
  }

  render() {
    const { intl, onChange, selectedColor } = this.props;
    const skinToneLabel = intl.formatMessage(messages.skinTone);
    const radioGroupStyle = { flexDirection: 'row' };
    const radioItemStyle = { padding: '2px' };

    return (
      <FormControl className="ColorSelect">
        <RadioGroup
          aria-label={skinToneLabel}
          name="skinTone"
          value={selectedColor}
          style={radioGroupStyle}
          onChange={onChange}
        >
          {this.state.skinToneMenu.map(skinTone => (
            <Radio
              key={skinTone.name}
              value={skinTone.color}
              style={radioItemStyle}
              icon={<Circle color={skinTone.color} />}
              checkedIcon={<Circle color={skinTone.color} />}
            />
          ))}
          {selectedColor && (
            <IconButton
              label={intl.formatMessage(messages.clearSelection)}
              onClick={() => {
                onChange();
              }}
            >
              <CloseIcon />
            </IconButton>
          )}
        </RadioGroup>
      </FormControl>
    );
  }
}

SkinToneSelect.propTypes = propTypes;
export default injectIntl(SkinToneSelect);
