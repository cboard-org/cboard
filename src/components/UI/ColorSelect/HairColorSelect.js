import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FormControl from '@material-ui/core/FormControl';
import { Card, CardContent, IconButton, Tooltip } from '@material-ui/core';
import { Face } from '@material-ui/icons';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';

import './ColorSelectDropdown.css';
import Circle from './Circle';
import messages from './HairColor.messages';

const hairColorSources = new Map([
  [
    'arasaac',
    [
      {
        name: 'blonde',
        color: '#fdd700'
      },
      {
        name: 'brown',
        color: '#a65e26'
      },
      {
        name: 'darkBrown',
        color: '#6a2703'
      },
      {
        name: 'gray',
        color: '#efefef'
      },
      {
        name: 'darkGray',
        color: '#aaabab'
      },
      {
        name: 'red',
        color: '#ed4120'
      },
      {
        name: 'black',
        color: '#020100'
      }
    ]
  ]
]);
const sourcesNames = new Map([['arasaac', 'ARASAAC']]);

const propTypes = {
  source: PropTypes.string.isRequired,
  intl: intlShape.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedColor: PropTypes.string.isRequired
};

class HairColorSelect extends React.Component {
  constructor(props) {
    super(props);

    this.wrapperRef = React.createRef();
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.state = {
      open: false,
      sourceName: sourcesNames.has(props.source)
        ? sourcesNames.get(props.source)
        : sourcesNames.get('arasaac'),
      hairColorMenu: hairColorSources.has(props.source)
        ? hairColorSources.get(props.source)
        : hairColorSources.get('arasaac')
    };
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  handleChange = event => {
    const { onChange } = this.props;

    onChange(event);
    this.setState({
      open: false
    });
  };

  handleClickOutside = event => {
    if (this.wrapperRef && !this.wrapperRef.current.contains(event.target)) {
      this.setState({
        open: false
      });
    }
  };

  toggleOpen() {
    this.setState({
      open: !this.state.open
    });
  }

  render() {
    const { intl, onChange, selectedColor, iconColor } = this.props;
    const hairColorLabel = `${this.state.sourceName} ${intl.formatMessage(
      messages.hairColor
    )}`;
    const radioGroupStyle = { flexDirection: 'column' };
    const radioItemStyle = { padding: '2px' };

    return (
      <div className="colorSelectDropdown" ref={this.wrapperRef}>
        <Tooltip title={hairColorLabel} aria-label={hairColorLabel}>
          <IconButton
            label={hairColorLabel}
            onClick={() => this.toggleOpen()}
            style={{ color: iconColor ? iconColor : 'inherit' }}
          >
            <Face />
          </IconButton>
        </Tooltip>
        <FormControl className="colorSelectDropdown-options">
          <Card className={this.state.open ? 'opened' : 'closed'}>
            <CardContent>
              <RadioGroup
                aria-label={hairColorLabel}
                name="hairColor"
                value={selectedColor}
                style={radioGroupStyle}
                onChange={this.handleChange}
              >
                {this.state.hairColorMenu.map(hairColor => (
                  <Radio
                    key={hairColor.name}
                    value={hairColor.name}
                    style={radioItemStyle}
                    icon={<Circle fill={hairColor.color} />}
                    checkedIcon={
                      <Circle
                        fill={hairColor.color}
                        color="primary"
                        strokeWidth={3}
                      />
                    }
                  />
                ))}
              </RadioGroup>
            </CardContent>
          </Card>
        </FormControl>
      </div>
    );
  }
}

HairColorSelect.propTypes = propTypes;
export default injectIntl(HairColorSelect);
