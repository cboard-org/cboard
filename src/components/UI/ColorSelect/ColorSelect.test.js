import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';

import ColorSelect from './ColorSelect';

describe('ColorSelect tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<ColorSelect />);
  });
});
