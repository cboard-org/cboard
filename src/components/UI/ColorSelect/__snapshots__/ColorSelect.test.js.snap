// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSelect tests default renderer 1`] = `
<ColorSelect
  colors={
    Array [
      "#CE93D8",
      "#2196F3",
      "#4CAF50",
      "#E57373",
    ]
  }
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
/>
`;
